spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/MiniBatch.scala

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.dataset
+
+import com.intel.analytics.bigdl.nn.abstractnn.Activity
+import com.intel.analytics.bigdl.tensor.Tensor
+
+/**
+ * A interface for MiniBatch.
+ * A MiniBatch contains a few samples.
+ *
+ * @tparam T Numeric type
+ */
+trait MiniBatch[T] {
+  /**
+   * Get the number of samples in this MiniBatch
+   * @return size How many samples in this MiniBatch
+   */
+  def size(): Int
+
+  /**
+   * Copy a piece of MiniBatch to dest MiniBatch buffer.
+   *
+   * Optimizer will hold a MiniBatch Buffer, this function will slice this MiniBatch to
+   * a smaller MiniBatch with offset and length, and copy to the MiniBatch Buffer.
+   *
+   * Notice: Be careful to avoid a heavy GC, or will lead to a bad performance. If the
+   * destIndexth in dest array is null, please create a new MiniBatch object and
+   * assigned this object to the dest array.
+   *
+   * @param offset offset, counted from 1
+   * @param length length
+   * @param dest Array of miniBatch
+   * @param destIndex array index
+   * @return A smaller MiniBatch
+   */
+  def copyTo(offset: Int, length: Int, dest: Array[MiniBatch[T]], destIndex: Int): Unit
```

Comment:
Instead of copyTo, provide a slice method instead; the user of this method can decide how to deal with the result.